### ISSUE TITLE
The Angular Way not showing 'Loading...' like in ajax way

### ISSUE THREAD

**Issue by shaniqwa** [l-lin/angular-datatables#1055]
## What versions you are using?

- node version: 6.11.0
- npm version: 3.10.10
- angular version: ^4.2.4
- jquery version: 3.2.1
- datatables version:
- angular-datatables version: ^4.1.1
- angular-cli version: 1.2.0
 -datatables.net: ^1.10.15

## What's the problem?
Hi there, first I want to say I have been using DataTables for a while in different projects and frameworks and its grate :)

I'm now using angular for my project, and I noticed there is a different behavior if using ajax or the angular way. 
I would much prefer to use the angular way, but I can't get the auto loader to show, while if using ajax to get the table content it shows up.
I have to implement my own loader before the whole table is shown. if not, I can see the table header not yet rendered and it looks buggy. 

Can this be fixed ? 
plus another question - whats the best way to implement different languages using angular ?

thanks 
Shani

**Comment by l-lin** [comment]
TLDR: You can't.

There are two loading with angular-datatables:

- One comes from your own loader. It is displayed before the table is rendered. Angular-datables offers the possibility to render the table at any time using `this.dtTrigger.next();`. So before rendering the table, the promises need to be resolved (or observable depending on what you use), thus adding a loading message will let users know that something is processing.
- The other comes from DataTables. The message Loading is displayed inside the Table while fetching the data from the server.

The second loader is from DataTables and angular-datatables cannot override the native function.

**Comment by shaniqwa** [comment]
I'm trying to user dtTrigger, but its not rendering the table at all. 
Here's my code: 
**Component**
```
@Component({
  selector: 'app-users',
  templateUrl: './users.component.html',
  styleUrls: ['./users.component.scss']
})

export class PageUsersComponent implements OnInit  {
  users: User[] = [];
  dtTrigger: Subject<any> = new Subject();
  dtOptions: any = {};
  displayUsers: boolean = false;
  error: boolean = false;

  public UsersActionConfig: any;

  constructor(private usersService: UsersService,
              public dialog: MdDialog,
              private router: Router) {}

  ngOnInit() {
    this.dtTrigger.next();
    this.UsersActionConfig = USERACTIONCONFIG;
    this.dtOptions = {
      // language: {
      //   url: "//cdn.datatables.net/plug-ins/9dcbecd42ad/i18n/Hebrew.json"
      // },
      dom: 'Bfrtip',
      lengthChange: true,
      buttons: [
        {extend: 'excel' , text:'<span mdTooltip="Export to Excel" mdTooltipPosition="above"><i class="material-icons">file_download</i></span>'},
        {extend: 'print' , text:'<i mdTooltip="Print" mdTooltipPosition="above" class="material-icons">print</i>'}
        ],
      iDisplayLength: 50,
    };
    this.usersService
      .getAllUsers()
      .subscribe((users) => {
          this.users = users;
          // Calling the DT trigger to manually render the table
          this.dtTrigger.next();
          this.displayUsers = true;
        },
        (error) => {
          this.displayUsers = true;
          this.error = true;
          this.dtTrigger.next();
        }
      );
  }
```
**html**
```
<div *ngIf="error;">
      <div class="callout callout-danger">
        <h4>Error Loading Data</h4>
        <p>Lorem ipsum dolor sit amet, consectetur adipisicing elit. Assumenda, alias, in accusantium totam adipisci vel et suscipit quidem libero pariatur minus ratione quo doloremque error at nemo incidunt dicta quia?</p>
      </div>
    </div>
    <ng-template #loading class="loaderHolder"><md-progress-spinner mode="indeterminate" style="margin: 0 auto">Loading...</md-progress-spinner></ng-template>
    <div *ngIf="displayUsers; else loading">
      <table datatable [dtOptions]="dtOptions" [dtTrigger]="dtTrigger" id="usersTable" class="row-border hover">
        <thead>
        <tr>
          <th>#</th>
          <th i18n="@@firstName">First name</th>
          <th i18n="@@lastName">Last name</th>
          <th i18n="@@phone">Phone</th>
          <th i18n="@@email">Email</th>
          <th i18n="@@wristId">Wrist ID</th>

        </tr>
        </thead>
        <tbody>
        <tr *ngFor="let user of users">
          <td>{{ user.id }}</td>
          <td>{{ user.firstName }}</td>
          <td>{{ user.lastName }}</td>
          <td>{{ user.phone }}</td>
          <td>{{ user.email }}</td>
          <td>{{ user.wristID }}</td>
        </tr>
        </tbody>
      </table>
```

first of all - can you tell whats wrong ? why it's not triggering ?
and second - is it possible to set ` this.displayUsers` to `true` only when rendering is done? meaning, does `this.dtTrigger.next()` has a callback ?

Thanks 
Shani

**Comment by l-lin** [comment]
It seems to be a timing issue. my guess is that `this.dtTrigger.next()` is called before [angular-datatable directive subscribes to the observable](https://github.com/l-lin/angular-datatables/blob/master/src/angular-datatables.directive.ts#L40).

As a workaround, you can use `setTimeout` as a "hack" it:

```typescript
ngOnInit() {
    this.UsersActionConfig = USERACTIONCONFIG;
    this.dtOptions = {
      // language: {
      //   url: "//cdn.datatables.net/plug-ins/9dcbecd42ad/i18n/Hebrew.json"
      // },
      dom: 'Bfrtip',
      lengthChange: true,
      buttons: [
        {extend: 'excel' , text:'<span mdTooltip="Export to Excel" mdTooltipPosition="above"><i class="material-icons">file_download</i></span>'},
        {extend: 'print' , text:'<i mdTooltip="Print" mdTooltipPosition="above" class="material-icons">print</i>'}
        ],
      iDisplayLength: 50,
    };
    this.usersService
      .getAllUsers()
      .subscribe((users) => {
          this.users = users;
          // Calling the DT trigger to manually render the table
          this.dtTrigger.next();
          setTimeout(() => {
            this.displayUsers = true;
          });
        },
        (error) => {
          this.displayUsers = true;
          this.error = true;
          setTimeout(() => {
            this.displayUsers = true;
          });
        }
      );
  }
```

**Comment by shaniqwa** [comment]
the problem is that is i'm adding ` [dtTrigger]="dtTrigger" ` to the html, then the table is not rendering to a datatable at all. i'ts just a normal table. I don't mind using a small timeout until it's finished to avoid buggy looking behavior, but I can't use the trigger at all at the moment.

**Comment by l-lin** [comment]
It's what [angular-datables](https://github.com/l-lin/angular-datatables/blob/v4.1.1/src/angular-datatables.directive.ts#L39-L42) has been coded for. If there is a `dtTrigger`, then the datatable won't render until the observable has been triggered.

**Comment by shaniqwa** [comment]
Yes, but even after it has been triggered the table is not rendered. 
Unfortunately there come a point when I must re-render :) when data changes ... I think its configuration/installation issues or some conflict preventing the trigger. Please any thoughts of how to debug this would be much appreciated.

**Comment by shaniqwa** [comment]
well I found that `dtTrigger` didn't work for me because the table was inside a `ngIf` block : 
`<div *ngIf="displayUsers; else loading">`
I tried putting `this.dtTrigger.next()` inside a timeout and boom - it worked. Then after trying few things I removed the template loader and now it works without any timeout. closing this issue, thanks

**Comment by ajmeersha** [comment]
@shaniqwa can u please share the code.. i stuck with the same problem
